### ISSUE TITLE
Cannot obtain a flat (terrain based) sparse point cloud

### ISSUE THREAD

**Issue by charlescva** [sweeneychris/TheiaSfM#131]
I am trying to create a sparse point cloud using images from video, with the default flags file for build_reconstruction.  I have had good success with a single pass around objects in my house using my smart phone, so I decided to try from an aerial platform like the Phantom DJI drone.

There are some variable zoom changes and not much in regard to camera intrinsics using video, so I've tried a few different optimizations such as FOCAL_LENGTH and PRINCIPAL_POINTS without much luck. It is not a fisheye, so RADIAL_DISTORTION is also not an issue.

I keep getting a sparse cloud that you can clearly see some 'flat' terrain, but it looks like the camera poses are off and rotated completely in the wrong direction a lot of the time.  The input set was about 2000 images and there were many matches.

In the video below you can see a rotation of the sparse cloud spun 180° with "None" for optimizations, as that seemed to yield the smoothest camera positions as the drone moved along a curve.  the spiral looking bits (upside down) are the camera poses.  But most importantly you can see clusters of points that create flat planes that are not properly aligned.

Any thoughts on how I can encourage theia to understand that the camera has changes in zoom and everything in the image is on the ground?  The altitude is only a few hundered meters, and the ground is generally flat.  If camera intrinsics like focal length are 100% required to do this, that would be good to know as well.

Thanks for your time.

 [Video of Sparse Point Cloud](https://youtu.be/_cJZSs-gyd8)

**Comment by charlescva** [comment]
So if i use less images on a single area without changing the distance, it seems to pose them correctly. It seems like the gradual variation in distance from lens to ground starts to throw it off as objects scale larger, resulting it many geometric planes being rotated about various points of origin.  When we talk about focal length in theia are we talking about the camera specifically, or is the real goal to specifcy distance to features? On an EO camera recording video, without auto focus enabled snapping photos every second that include exif data, we tend to have a static focal point regardless of camera pose.  Is there a setting which would encourage posing along a similar geometric plane, like the ground? A sort of maximum delta in which the jacobian manipulation is applied? Sorry if this is coming out like giberish but im still learning a lot.

Thanks

**Comment by charlescva** [comment]
Got a better comparison...
![comparison](https://cloud.githubusercontent.com/assets/5343462/21774463/a142fbb0-d660-11e6-8ffb-c908917070da.png)

Notice how the peaks of the houses are misaligned.  Any tips?

**Comment by sweeneychris** [comment]
A few things here:

1) Radial distortion has to do with lens distortion -- it has nothing to do with fisheye cameras specifically. The DJI drones I have used in the past do indeed have a small amount of lens distortion.

2) Zoom in the camera means there is a change in focal length. I would highly discourage this, because if you have a consistent focal length throughout the video sequence then you can turn --shared_calibration on in the parameter flags. This should give you better intrinsics.

3) I would try the incremental pipeline instead of just the global pipeline. Incremental tends to me more robust.

4) There is an unsolvable ambiguity with the ground plane and radial distortion for aerial captures like this. This is not a limitation of Theia, but simply an ill-posed problem: http://www.cv-foundation.org/openaccess/content_cvpr_2014/papers/Wu_Critical_Configurations_For_2014_CVPR_paper.pdf

**Comment by charlescva** [comment]
This is very helpful information Chris, thank you!  After reviewing several reconstructions, I believe there is either FOCAL_LENGTH or RADIAL_DISTORTION or both at play here.  Also getting the camera metadata is key.  I'm thinking the smart thing to do will be to inject the EXIF data into the JPEGs based on the specification of the video. e.g. Focal Length.  Havent tried it yet but this tool: http://www.sno.phy.queensu.ca/~phil/exiftool/ seems promising.  

Been looking at matlab calibration toolkits, but in some cases I may not have physical access to the original lens/camera to create a distortion model and correct the images before hand.  After thinking this over 24 hours, I can see precisely why this is definitely an 'ill-posed' problem!

I'm going to continue working at this until I find a viable solution.  On a side not I've containerized (docker) theia and have forked this repo.  The combination of BTRFS and physical CPU allocation via kernel space gives me good flexibility and great performance.  Would you be interested in a Dockerfile being contributed to your Docs?  This would perform an install of eigen, and a full compile of oiio, and then theia from source/latest.  Thoughts?

**Comment by sweeneychris** [comment]
To your first point, if you simply set the CameraIntrinsicsPrior parameters to the estimated (or offline calibrated) focal and radial distortion then Theia will use that as a starting point. Otherwise it will try to get the EXIF focal length, or make other intelligent guesses.

I don't know anything about Docker really, but multiple people have asked about Theia + Docker so I think it would definitely be useful if you could add the Dockerfile and whatever else is necessary (documentation on the website?) for others to reproduce the desired Docker setup. Sounds awesome!

**Comment by charlescva** [comment]
Had great success last night when I utilized a higher resolution images at 1.98 MP each.  (Doh!)  I had been using a lower resolution at around 0.88 MP when I got the previous results due to some mistakes on my end between video capture and image extraction.  With that being said, it looks amazing! Chris, I can truly say this is one of the most amazing projects I've ever had the privilege to work on.

![neighborhood](https://cloud.githubusercontent.com/assets/5343462/21846850/fec4aba0-d7c6-11e6-964f-aba4eb20c4dd.png)
[https://youtu.be/We_MaePhOmM](Blender Video)

Going to spend some time today formalizing the docker integration and create a pull request to add to the the build portion of the theia docs.  I think that would be a good home for the dockerfile right now.  If it gains traction I can always help with publishing official containers on dockerhub, and would be happy to do it with 0.7.

Thanks again!

**Comment by sweeneychris** [comment]
Hey that looks pretty awesome! Glad that bumping up the image resolution helped. Higher resolution is always better assuming you can wait for the increased runtime. By the way, one "gotcha" I've run into previously with extracting images from video is not extracting full-res images. With ffmpeg and avconv you can adjust the output image quality with the --qscale command line flag.

**Comment by charlescva** [comment]
Yes! That was definitely a issue I hit earlier, qscale had between 2-32, with a default that is supposedly undetectable by human eye.  switched to '2' when I noticed a few artifacts that were not in the original video and did some investigation on the jpeg defaults.

The -r flag will extract at a rate of frame/second.  so 10/1 would be 10 Jpegs/second, while the example below is 1 image per second.

e.g.
`ffmpeg -i ~/Videos/Video.mp4 -qscale 2 -r 1/1 output/%03d.jpg`

I've noticed the density of the sparse cloud can be significantly increased by providing more frame/second input jpegs. Thus, 10/1 will generate more vertices than 1/1.
